amazonsns/publisher.go

```diff
@@ -0,0 +1,66 @@
+package amazonsns
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sns"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+
+	watermill_amazonsqs "github.com/ThreeDotsLabs/watermill-amazonsqs"
+	"github.com/ThreeDotsLabs/watermill-amazonsqs/amazonsqs"
+)
+
+type Publisher struct {
+	config watermill_amazonsqs.PublisherConfig
+	logger watermill.LoggerAdapter
+	sns    *sns.SNS
+}
+
+func NewPublisher(config watermill_amazonsqs.PublisherConfig, logger watermill.LoggerAdapter) (*Publisher, error) {
+	config.AWSConfig = amazonsqs.SetEndPoint(config.AWSConfig)
+	sess, err := session.NewSession(&config.AWSConfig)
+	if err != nil {
+		// TODO wrap
+		return nil, err
+	}
+
+	return &Publisher{
+		sns:    sns.New(sess),
+		config: config,
+		logger: logger,
+	}, nil
+}
+
+func (p Publisher) Publish(topic string, messages ...*message.Message) error {
+	// TODO method for generating
+	topicInfo, err := p.sns.CreateTopic(&sns.CreateTopicInput{
```

Comment:
Do we need to check if the topic exists?

Reply:
The "CreateTopic" will create it if it does not exist the problem as I stated is if you publish to a topic that does not have subscriptions the messages will be lost.  Normally enterprises would create the topic/queues/endpoints and subscriptions (maybe using terraform (or the like) and then use them in a watermill flow.

Reply:
Won't it crash if the topic already exists though?

Reply:
Nope in that case the topic info is returned to the called to include the TopicARN that you need in the Publish call. https://docs.aws.amazon.com/sns/latest/api/API_CreateTopic.html